File: quattro2/win10/Network/HTTPConnection.cpp
/*
 * @(#)HTTPConnection.cpp
 *
 * Copyright 2015 Roland Corporation, Japan. All rights reserved.
 */

#include "HTTPConnection.h"
#include "HTTPDownload.h"
#include "HTTPUpload.h"

HTTPConnection::~HTTPConnection()
{
	{
		mutex_lock mtx(mtx_map);
		std::map<DWORD, HTTPTask*>::iterator iter = idMap.begin();
		for ( ; iter != idMap.end(); iter++) {
			HTTPTask* obj = iter->second;
			obj->abort();
		}
	}

	while (!idMap.empty()) ;
}

DWORD HTTPConnection::download(LPCTSTR url, LPCTSTR file)
{
	if (url == NULL)
		return 0xffffffff;

	return (new HTTPDownload(this, url, file))->_threadStart(true);
}

DWORD HTTPConnection::upload(LPCTSTR url, LPCTSTR file)
{
	if (url == NULL || file == NULL)
		return 0xffffffff;

	return (new HTTPUpload(this, url, file))->_threadStart(true);
}

void HTTPConnection::cancel(DWORD id)
{
	mutex_lock mtx(mtx_map);

	std::map<DWORD, HTTPTask*>::iterator iter = idMap.find(id);
	if (iter != idMap.end()) {
		HTTPTask* obj = iter->second;
		obj->abort();
	}
}

void HTTPConnection::insert(DWORD id, HTTPTask* obj)
{
	mutex_lock mtx(mtx_map);

	idMap.insert(std::make_pair(id, obj));
}

void HTTPConnection::remove(DWORD id)
{
	mutex_lock mtx(mtx_map);

	std::map<DWORD, HTTPTask*>::iterator iter = idMap.find(id);
	if (iter != idMap.end()) {
		idMap.erase(iter);
	}
}

void HTTPConnection::progress(DWORD id, DWORD contentLength, DWORD loaded)
{
	if (mp_delegate) {
		mp_delegate->connectionDidLoadData(id, contentLength, loaded);
	}
}

void HTTPConnection::finish(DWORD id, LPCTSTR file)
{
	if (mp_delegate) {
		mp_delegate->connectionDidFinishLoading(id, file);
	}
}

void HTTPConnection::error(DWORD id, LPCTSTR url)
{
	if (mp_delegate) {
		mp_delegate->connectionErrorDidOccur(id, url);
	}
}

File: quattro2/android/main/java/quattro/Network/HTTPConnection.java
//
//	HTTPConnection.java
//
//	Copyright 2015 Roland Corporation. All rights reserved.
//

package quattro.Network;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.HashMap;

public class HTTPConnection {

	public HTTPConnectionDelegate delegate = null;

	private Context applicationContext;
	private int taskId = 0;
	private HashMap<String, HTTPTask> tasks = new HashMap<String, HTTPTask>();

	public HTTPConnection(Context c) {
		applicationContext = c;
	}

	public int download(URL url, File file) {
		if (url == null)
			return -1;
		if (file == null) {
			try {
				file = File.createTempFile("___", null, applicationContext.getExternalCacheDir());
			} catch (IOException e) { return -1; }
		}

		HTTPTask task = new HTTPDownload(this);
		synchronized (tasks) {
			task.id = taskId++;
			tasks.put(String.valueOf(task.id), task);
		}
		task.url = url;
		task.file = file;
		task.execute();
		return task.id;
	}

	public int upload(URL url, File file) {
		if (url == null || file == null)
			return -1;

		HTTPTask task = new HTTPUpload(this);
		synchronized (tasks) {
			task.id = taskId++;
			tasks.put(String.valueOf(task.id), task);
		}
		task.url = url;
		task.file = file;
		task.execute();
		return task.id;
	}

	public void cancel(int id) {
		synchronized (tasks) {
			HTTPTask task = tasks.get(String.valueOf(id));
			if (task != null) {
				task.cancel(true);
				tasks.remove(String.valueOf(id));
			}
		}
	}

	protected void progress(HTTPTask obj, int contentLength, int amount) {
		if (delegate != null) {
			delegate.connectionDidLoadData(obj.id, contentLength, amount);
		}
	}

	protected void finish(HTTPTask obj, File file) {
		if (delegate != null) {
			delegate.connectionDidFinishLoading(obj.id, obj.file);
		}
		synchronized (tasks) {
			tasks.remove(String.valueOf(obj.id));
		}
	}

	protected void error(HTTPTask obj, URL url) {
		if (delegate != null) {
			delegate.connectionErrorDidOccur(obj.id, obj.url);
		}
		synchronized (tasks) {
			tasks.remove(String.valueOf(obj.id));
		}
	}

	public void destroy() {
		synchronized (tasks) {
			for (HTTPTask task : tasks.values()) {
				task.cancel(true);
			}
		}
		delegate = null;
		applicationContext = null;
	}

}
